### ISSUE TITLE
Order of dimensions in multichannel images

### ISSUE THREAD

**Issue by caspervdw** [soft-matter/pims#124]
In PR https://github.com/soft-matter/pims/pull/118 and https://github.com/soft-matter/pims/pull/109, I started working on importing images with more than one channel (C), focal plane (Z) and time points (T). The dimensional order I've adopted is (TCZYX), starting with the outer dimension.

I've read https://github.com/scikit-image/scikit-image/pull/1280 and the scikit convention is to put multichannels in the last dimension. I have a strong tendency to go along with this convention, but it does raise a few issues:

- this dimensional order is unconventional in microscopy (except for RGB). For example, ImageJ, which is the most standard image analysis tools always has YX as last two dimensions.
- `trackpy` operates on grayscale images, so a conversion `process_func` is always necessary if you import a multichannel. For instance, subtracting two channels is less computer-intensive when channel is the first dimension. (compare `image[0] - image[1]` to `image[:,:,0] - image[:,:,1]`)
- `_repr_html_` in `frame.py` distinguishes multichannels from RGBs based on the location of multichannel axis. A 3-channel multichannel would be indistinguishable from RGB, unless we decide to give the `frame` object an `isRGB` property. Which I am in favour of.

Question is: do we keep it as it is or should we go along with scikit-image and put channels in the inner dimension? What is your opinion?

**Comment by caspervdw** [comment]
PR 109 and PR 118 are sort of waiting for this decision. What do you think about this @tacaswell @danielballan?

**Comment by danielballan** [comment]
Hey, sorry for the delay, first week working at BNL with @tacaswell! Reviewing now....

**Comment by tacaswell** [comment]
Sorry, @danielballan has been busy defending and moving in the past month, I don't have as good of an excuse.

On further consideration, the skimage conventions are for storing bunches of images together in memory, where as pims is more about getting images stored in files into memory with out having to special-case file formats everywhere.  The file formats that drove Dan and I to work on this were derived from what we had to deal with, which were single streams of gray-scale or RGB frames, which leaks into the assumptions made by pims.

To that end, I think we can be flexible about this sort of ordering.  The sub-classes which _know_ about channels should be able to be configured at runtime as to the order they loop over or aggregate the raw frames.  Maybe something like:

```python
class ChannelZAware(self, ...., loop_order=['z', 'c', 't']):
     pass
```

so when you do `for im in cz_awar` it loops over all z at fixed (c, t) etc.  I don't claim this is the best scheme for notating this.

I suspect you could also do something similar for aggrecation

```python
class ChannelZAware(sellf, ..., aggregate='xyz', loop_order=['t', 'c']):
    pass
```

which would iterate over XYZ blocks over time at a fixed channel and then over channels.

**Comment by caspervdw** [comment]
Being flexible about this would be great, nice ideas. We could implement this in a class `Multidimensional` that defines `get_frame` and some necessary properties (`loop_order`, `aggregate`, `count`, `channel`). Readers `ImageSequence3D` and `Bioformats3D` are then derived from `(FramesSequence, Multidimensional)`

I could make this class with the current behaviour of `ImageSequence3D` and then the flexibility can be added at a later stage.

**Comment by caspervdw** [comment]
Solved as of PR #151
